src/parallel-rustc.md

```diff
@@ -1,34 +1,111 @@
+As of <!-- date-check --> August 2022, the only stage of the compiler that 
+is already parallel is codegen. Some other parts of the nightly compiler 
+have parallel implementations, such as query evaluation, type check and 
+monomorphization, but there is still a lot of work to be done. The lack of 
+parallelism at other stages (for example, macro expansion) also represents 
+an opportunity for improving compiler performance.
+
+**To try out the current parallel compiler**, one can install rustc from 
+source code with enable `parallel-compiler = true` in the `config.toml`.
 
 These next few sections describe where and how parallelism is currently used,
 and the current status of making parallel compilation the default in `rustc`.
 
-The underlying thread-safe data-structures used in the parallel compiler
-can be found in the `rustc_data_structures::sync` module. Some of these data structures
-use the `parking_lot` crate as well.
-
 ## Codegen
 
-There are two underlying thread safe data structures used in code generation:
-
-- `Lrc`
-    -  Which is an [`Arc`][Arc] if `parallel_compiler` is true, and a [`Rc`][Rc]
-       if it is not.
-- `MetadataRef` -> [`OwningRef<Box<dyn Erased + Send + Sync>, [u8]>`][OwningRef]
-    - This data structure is specific to `rustc`.
-
 During [monomorphization][monomorphization] the compiler splits up all the code to
 be generated into smaller chunks called _codegen units_. These are then generated by
 independent instances of LLVM running in parallel. At the end, the linker
 is run to combine all the codegen units together into one binary. This process
 occurs in the `rustc_codegen_ssa::base` module.
 
+## Data Structures
+
+The underlying thread-safe data-structures used in the parallel compiler
+can be found in the `rustc_data_structures::sync` module. These data structures 
+are implemented diferently depending on whether `parallel-compiler` is true.
+
+| data structure                   | parallel                                            | non-parallel |
+| -------------------------------- | --------------------------------------------------- | ------------ |
+| Lrc                              | std::sync::Arc | std::rc::Rc |
+| Weak                             | std::sync::Weak                                     | std::rc::Weak |
+| Atomic{Bool}/{Usize}/{U32}/{U64} | std::sync::atomic::Atomic{Bool}/{Usize}/{U32}/{U64} | (Cell\<Bool/Usize/U32/U64>) |
+| OnceCell                         | std::sync::OnceLock                                 | std::cell::OnceCell |
+| Lock\<T> | (parking_lot::Mutex\<T>) | (std::cell::RefCell) |
+| RwLock\<T> | (parking_lot::RwLock\<T>) | (std::cell::RefCell) |
+| MTRef<'a, T> | &'a T | &'a mut T |
+| MTLock\<T> | (Lock\<T>) | (T) |
+| ReadGuard | parking_lot::RwLockReadGuard | std::cell::Ref |
+| MappedReadGuard | parking_lot::MappedRwLockReadGuard | std::cell::Ref |
+| WriteGuard | MappedWriteGuard | std::cell::RefMut |
```

Comment:
```suggestion
| WriteGuard | parking_lot::MappedRwLockWriteGuard | std::cell::RefMut |
```
it is more easy to refer to original type, especially since alias is not yet shown in this table